csrc/src/lib.rs

```diff
@@ -1403,6 +1436,706 @@ pub fn flash_attn_varlen_full(
     q.apply_op3(k, v, op)
 }
 
+/// Flash-attention v2 layer, with Key-Value cache.
+///
+/// NOTE: We are not passing in each Key and Value tensor for the decoding phase. This is
+/// because we plan to use paged attention with flash attention.
+/// In that case, the key and value tensors at each decoding phase are stored within the
+/// kv cache tensor, separately. So we don't need to pass the key and value tensors to the
+/// flash attention kernel, directly.
+struct FlashAttentionKvCache {
+    /// Softmax scale
+    pub softmax_scale: f32,
+    /// Block table, used for paged attention algorithm
+    /// of shape [batch_size, max_num_block_per_sequence]
+    pub block_table: Option<Tensor>,
+    /// Alibi slopes, see https://nn.labml.ai/transformers/alibi/index.html,
+    /// of shape `[num_heads, ]` or `[batch_size, num_heads]`
+    pub alibi_slopes: Option<Tensor>,
+    /// Window size for left sided slicing attention
+    pub window_size_left: Option<usize>,
+    /// Window size for right sided slicing attention
+    pub window_size_right: Option<usize>,
+    /// Sequence lengths for the key tensor,
+    /// of shape `[batch_size, ]`
+    pub seqlens_k: Option<Tensor>,
+    /// Softcap parameter, used in Grok and Gemma2 models
+    pub softcap: Option<f32>,
+}
+
+impl FlashAttentionKvCache {
+    fn cuda_fwd_t<
```

Comment:
This function has 500+ lines, it would be nice to split it, but I understand if it's not optimal for split.

Reply:
Can we leave it as is, for now ?